dhall/src/Dhall/Format.hs

```diff
@@ -117,3 +131,100 @@ format (Format { input = input0, ..}) = go input0
                         let modified = "formatted"
 
                         Control.Exception.throwIO CheckFailed{..}
+
+--
+detectCharacterSet :: Expr Src a -> CharacterSet
+detectCharacterSet = go mempty
+  where
+    recurse = go ASCII
+
+    -- Short-circuit traversal if Unicode encountered
+    go Unicode = \_ -> Unicode
+    go ASCII = \case
+        Note Src{srcText} expr ->
+            case expr of
+                Lam _ _
+                    | "λ" `Text.isInfixOf` srcText ||
+                      "→" `Text.isInfixOf` srcText -> Unicode
+                Pi _ _ _
+                    | "∀" `Text.isInfixOf` srcText ||
+                      "→" `Text.isInfixOf` srcText -> Unicode
+                Prefer _ _ _
+                    | "⫽" `Text.isInfixOf` srcText -> Unicode
+                Combine _ _ _
+                    | "∧" `Text.isInfixOf` srcText -> Unicode
+                CombineTypes _ _
+                    | "⩓" `Text.isInfixOf` srcText -> Unicode
+                _ -> recurse expr
```

Comment:
I think we might still get some false positives, for example if one of these Unicode characters appears in a string or comment within these specific source spans.

I'm not sure how to improve this though. The only other approach I have in mind involve changing the `Parser` type, which seems a bit heavy handed :sweat_smile:

Reply:
Well I guess I could add an extra field for `CharacterSet` to `Lam`, `Pi`, `Prefer`, `Combine`, and `CombineTypes`. That might actually be the cleanest option.

Reply:
Yeah, I think the reliable solution is to extend those constructors to store which character set was parsed

Reply:
I have now implemented to reliable solution. It makes the PR quite a bit bigger, but at least we won't get any weird behaviour around formatting.

I hope this won't affect parsing performance. Do we have any benchmarks in place?

Reply:
I found some benchmarks! :smile:

Reply:
So, this more complete solution kind of makes equality for `Expr` a bit more annoying. I am currently not propagating the character set into the evaluation types, which means I lose that information when not dealing with actual source code. I thought that would be a fine approach except it breaks all the tests that rely on equality not caring about character set (which is many of them :sweat_smile: )

Would a hacky `Eq` instance for `CharacterSet` where `_ == _ = True` be acceptable? :stuck_out_tongue_closed_eyes:

Reply:
Or maybe I should take a similar approach to what we do for the imports. I could have an extra type level argument for `Expr` and then `Void` out that field when the `Expr` doesn't represent something that was directly inputted by text.

Reply:
Could you give an example of a test that relies on equality on `Expr`?

Reply:
There's quite a few of them (around 200): https://hydra.dhall-lang.org/build/79114/nixlog/2

I gave the extra parameter to `Expr` a go. It unfortunately involves a lot of code changes. And I don't think it can work. Maybe of the traversal code would require the extra parameter to be a `Monoid` and `Void` is only a `Semigroup`.

One other alternative to fix this is to actually properly propagate `CharacterSet` through the evaluation (in types like `Val`). I will try this approach.

Reply:
> One other alternative to fix this is to actually properly propagate CharacterSet through the evaluation (in types like Val). I will try this approach.

This does fix many of the failing tests but not all. Some of them seem unfixable with this approach. For example binary serialisation loses that information (as it should). And typechecking now fails to unify if it inferred different `CharacterSet`s, which is obviously very silly :sweat_smile:

At the moment I think either I need to revert to the `Src` based solution which wasn't very robust since it could falsely detect unicode when it wasn't actually used in syntax (badly placed comments), or I need to add that hacky `Eq` instance. I could potentially but that instance on a newtype to make it more explicit:

```haskell
newtype AlwaysEq a = AlwaysEq a

instance Eq (AlwaysEq a) where
    _ == _ = True
```

And then use `AlwaysEq CharacterSet` in the `Expr` AST. It's still a bit ugly but at least there shouldn't be too many surprises.

Reply:
Actually, what I think you can do is to make the field a `Maybe` value, and parsing provides a `Just` value and `denote` turns it into `Nothing`.  This would be consistent with how we handle other source spans in the code.

Reply:
Right! That actually makes more sense. I'm not sure why I didn't think that Src would also be an issue for Eq.

I will do that thanks! 👍

Reply:
I implemented your `Maybe CharacterSet` solution, much nicer than what I was attempting! Thanks for the suggestion :smile: